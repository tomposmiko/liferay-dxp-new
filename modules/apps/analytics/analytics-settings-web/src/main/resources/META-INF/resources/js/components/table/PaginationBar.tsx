/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

import ClayButton from '@clayui/button';
import ClayIcon from '@clayui/icon';
import {ClayPaginationWithBasicItems} from '@clayui/pagination';
import ClayPaginationBar from '@clayui/pagination-bar';
import {sub} from 'frontend-js-web';
import React from 'react';

import {DELTAS, TPagination} from '../../utils/pagination';
import {Events, useData, useDispatch} from './Context';

interface IPaginationBarProps {
	disabled: boolean;
}

const PaginationBar: React.FC<IPaginationBarProps> = ({disabled}) => {
	const {
		pagination: {page, pageSize, totalCount},
	} = useData();
	const dispatch = useDispatch();

	if (!totalCount || disabled) {
		return null;
	}

	return (
		<ClayPaginationBar>
			<ClayPaginationBar.DropDown
				items={DELTAS.map((pageSize: TPagination['pageSize']) => ({
					label: String(pageSize),
					onClick: () => {
						dispatch({
							payload: {
								page: 1,
								pageSize,
							},
							type: Events.ChangePagination,
						});
					},
				}))}
				trigger={
					<ClayButton displayType="unstyled">
						<strong>
							{sub(Liferay.Language.get('x-entries'), pageSize)}
						</strong>

						<ClayIcon symbol="caret-double-l" />
					</ClayButton>
				}
			/>

			<ClayPaginationBar.Results>
				{sub(Liferay.Language.get('showing-x-to-x-of-x-entries'), [
					(page - 1) * pageSize + 1,
					page * pageSize < totalCount ? page * pageSize : totalCount,
					totalCount,
				])}
			</ClayPaginationBar.Results>

			<ClayPaginationWithBasicItems
				active={page}
				defaultActive={1}
				onActiveChange={(page: number) => {
					dispatch({
						payload: {page},
						type: Events.ChangePagination,
					});
				}}
				totalPages={Math.ceil(totalCount / pageSize)}
			/>
		</ClayPaginationBar>
	);
};

export default PaginationBar;
