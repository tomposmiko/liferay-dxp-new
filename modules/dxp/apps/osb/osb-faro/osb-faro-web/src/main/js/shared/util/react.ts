import {isEqual} from 'lodash';

export const getDisplayName = WrappedComponent =>
	WrappedComponent.displayName || WrappedComponent.name || 'Component';

type HasChanges = <T>(
	prev: T | object,
	next: T | object,
	...keys: string[]
) => boolean;

/**
 * Compare previous state or props object by provided keys to detect changes.
 */
export const hasChanges: HasChanges = (prev = {}, next = {}, ...keys) => {
	for (const key of keys) {
		if (key in next) {
			const newVal = next[key];

			const prevVal = prev[key];

			if (!isEqual(newVal, prevVal)) {
				return true;
			}
		}
	}

	return false;
};
