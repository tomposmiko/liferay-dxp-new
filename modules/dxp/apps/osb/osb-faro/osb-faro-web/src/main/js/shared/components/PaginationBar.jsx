import autobind from 'autobind-decorator';
import ClayButton from '@clayui/button';
import ClayDropDown from '@clayui/drop-down';
import ClayIcon from '@clayui/icon';
import faroConstants from 'shared/util/constants';
import getCN from 'classnames';
import NavBar from './NavBar';
import Pagination from './Pagination';
import React from 'react';
import {getPluralMessage} from '../util/lang';
import {PropTypes} from 'prop-types';
import {setUriQueryValues} from 'shared/util/router';

const {cur: DEFAULT_CUR, deltaValues} = faroConstants.pagination;

const FIRST_PAGE = 1;

const MAXIMUM_DELTA = 75;

const SIZES = ['sm', 'lg'];

class DeltaItem extends React.Component {
	static propTypes = {
		delta: PropTypes.number.isRequired,
		onChange: PropTypes.func
	};

	@autobind
	handleChange() {
		const {delta, onChange} = this.props;

		if (onChange) {
			onChange(delta);
		}
	}

	render() {
		return (
			<ClayDropDown.Item {...this.props} onClick={this.handleChange}>
				{this.props.children}
			</ClayDropDown.Item>
		);
	}
}

class PaginationBar extends React.Component {
	static defaultProps = {
		deltas: deltaValues,
		selectedDelta: deltaValues[1],
		showDeltaDropdown: true,
		showResultsMessage: true,
		totalItems: 0
	};

	static propTypes = {
		deltas: PropTypes.array,
		href: PropTypes.string,
		onDeltaChange: PropTypes.func,
		onPageChange: PropTypes.func,
		page: PropTypes.number,
		selectedDelta: PropTypes.number,
		showDeltaDropdown: PropTypes.bool,
		showResultsMessage: PropTypes.bool,
		size: PropTypes.oneOf(SIZES),
		totalItems: PropTypes.number
	};

	state = {};

	getStartedItem(start, selectedDelta, totalItems) {
		let currentStart = this.startItem(start, selectedDelta);

		if (currentStart > totalItems) {
			currentStart = this.startItem(
				Math.ceil(totalItems / selectedDelta) * selectedDelta,
				selectedDelta
			);
		}

		return currentStart.toLocaleString();
	}

	startItem(start, selectedDelta) {
		return start - selectedDelta + 1;
	}

	render() {
		const {
			className,
			deltas,
			href,
			onDeltaChange,
			onPageChange,
			showDeltaDropdown,
			showResultsMessage,
			size,
			totalItems
		} = this.props;

		const classes = getCN('pagination-bar-root', className, {
			[`pagination-${size}`]: size
		});

		const page = Math.max(this.props.page, FIRST_PAGE) || FIRST_PAGE;

		const selectedDelta =
			Math.min(this.props.selectedDelta, MAXIMUM_DELTA) || MAXIMUM_DELTA;

		const start = page * selectedDelta;

		return (
			<NavBar className={classes}>
				{showDeltaDropdown && (
					<ClayDropDown
						className='dropdown-root pagination-items-per-page'
						closeOnClick
						trigger={
							<ClayButton
								className='button-root'
								displayType='secondary'
								size='sm'
							>
								{getPluralMessage(
									Liferay.Language.get('x-item'),
									Liferay.Language.get('x-items'),
									selectedDelta
								)}

								<ClayIcon
									className='ml-2'
									symbol='caret-bottom'
								/>
							</ClayButton>
						}
					>
						{deltas.map(item => (
							<DeltaItem
								delta={item}
								href={
									onDeltaChange
										? null
										: setUriQueryValues({
												delta: item,
												page: DEFAULT_CUR
										  })
								}
								key={item}
								onChange={onDeltaChange}
							>
								{item}
							</DeltaItem>
						))}
					</ClayDropDown>
				)}

				{showResultsMessage && (
					<div
						className='pagination-results'
						key='PAGINATION_RESULTS'
					>
						{getPluralMessage(
							Liferay.Language.get('showing-x-to-x-of-x-entry'),
							Liferay.Language.get('showing-x-to-x-of-x-entries'),
							totalItems,
							true,
							[
								this.getStartedItem(
									start,
									selectedDelta,
									totalItems
								),
								Math.min(start, totalItems).toLocaleString(),
								totalItems.toLocaleString()
							]
						)}
					</div>
				)}

				<Pagination
					href={href}
					onChange={onPageChange}
					page={
						page <= totalItems
							? page
							: Math.ceil(totalItems / selectedDelta)
					}
					total={Math.ceil(totalItems / selectedDelta)}
				/>
			</NavBar>
		);
	}
}

export default PaginationBar;
