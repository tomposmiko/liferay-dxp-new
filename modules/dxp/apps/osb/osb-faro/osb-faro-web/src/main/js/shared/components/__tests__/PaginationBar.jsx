import PaginationBar from '../PaginationBar';
import React from 'react';
import {cleanup, render} from '@testing-library/react';
import {StaticRouter} from 'react-router';

jest.unmock('react-dom');

const DefaultComponent = props => (
	<StaticRouter>
		<PaginationBar
			href=''
			page={3}
			selectedDelta={10}
			totalItems={100}
			{...props}
		/>
	</StaticRouter>
);

describe('PaginationBar', () => {
	afterEach(cleanup);

	it('should render', () => {
		const {container} = render(<DefaultComponent />);
		expect(container).toMatchSnapshot();
	});

	it('should render with small size', () => {
		const {container} = render(<DefaultComponent size='sm' />);

		expect(container.querySelector('.pagination-sm')).toBeTruthy();
	});

	it('should render with large size', () => {
		const {container} = render(<DefaultComponent size='lg' />);

		expect(container.querySelector('.pagination-lg')).toBeTruthy();
	});

	it('should render different deltas', () => {
		const {container} = render(
			<DefaultComponent
				deltas={[1, 2, 3, 4]}
				page={1}
				selectedDelta={1}
				totalItems={10}
			/>
		);

		expect(container).toMatchSnapshot();
	});
});
