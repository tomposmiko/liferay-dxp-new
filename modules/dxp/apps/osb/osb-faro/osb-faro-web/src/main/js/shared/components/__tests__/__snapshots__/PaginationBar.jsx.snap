// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaginationBar should render 1`] = `
<div>
  <nav
    class="navbar navbar-root pagination-bar-root"
  >
    <div
      class="container-fluid"
    >
      <div
        aria-haspopup="true"
        class="dropdown dropdown-root pagination-items-per-page"
      >
        <button
          class="button-root dropdown-toggle show-caret btn btn-sm btn-unstyled"
          type="button"
        >
          <span
            class="text-truncate"
          >
            10 Items
          </span>
          
          <span
            class="caret-root"
          >
            <svg
              class="lexicon-icon lexicon-icon-caret-bottom icon-root"
              role="presentation"
            >
              <use
                xlink:href="#caret-bottom"
              />
            </svg>
          </span>
        </button>
      </div>
      <div
        class="pagination-results"
      >
        Showing 21 to 30 of 100 entries.
      </div>
      <ul
        class="pagination pagination-root"
      >
        <li
          class="page-item"
        >
          <a
            class="button-root page-link btn btn-unstyled"
            href="/?page=2"
          >
            <svg
              class="lexicon-icon lexicon-icon-angle-left icon-root"
              role="presentation"
            >
              <use
                xlink:href="#angle-left"
              />
            </svg>
          </a>
        </li>
        <li
          class="page-item"
        >
          <a
            class="button-root page-link btn btn-unstyled"
            href="/?page=1"
          >
            1
          </a>
        </li>
        <li
          class="page-item"
        >
          <a
            class="button-root page-link btn btn-unstyled"
            href="/?page=2"
          >
            2
          </a>
        </li>
        <li
          class="page-item active"
        >
          <a
            class="button-root page-link btn btn-unstyled"
            href="/?page=3"
          >
            3
          </a>
        </li>
        <li
          class="page-item"
        >
          <a
            class="button-root page-link btn btn-unstyled"
            href="/?page=4"
          >
            4
          </a>
        </li>
        <li
          class="page-item"
        >
          <a
            class="button-root page-link btn btn-unstyled"
            href="/?page=5"
          >
            5
          </a>
        </li>
        <li
          class="page-item"
        >
          <div
            aria-haspopup="true"
            class="dropdown dropdown-root"
          >
            <button
              class="button-root dropdown-toggle page-link btn btn-unstyled"
              type="button"
            >
              <span
                class="text-truncate"
              >
                ...
              </span>
              
            </button>
          </div>
        </li>
        <li
          class="page-item"
        >
          <a
            class="button-root page-link btn btn-unstyled"
            href="/?page=10"
          >
            10
          </a>
        </li>
        <li
          class="page-item"
        >
          <a
            class="button-root page-link btn btn-unstyled"
            href="/?page=4"
          >
            <svg
              class="lexicon-icon lexicon-icon-angle-right icon-root"
              role="presentation"
            >
              <use
                xlink:href="#angle-right"
              />
            </svg>
          </a>
        </li>
      </ul>
    </div>
  </nav>
</div>
`;

exports[`PaginationBar should render different deltas 1`] = `
<div>
  <nav
    class="navbar navbar-root pagination-bar-root"
  >
    <div
      class="container-fluid"
    >
      <div
        aria-haspopup="true"
        class="dropdown dropdown-root pagination-items-per-page"
      >
        <button
          class="button-root dropdown-toggle show-caret btn btn-sm btn-unstyled"
          type="button"
        >
          <span
            class="text-truncate"
          >
            1 Item
          </span>
          
          <span
            class="caret-root"
          >
            <svg
              class="lexicon-icon lexicon-icon-caret-bottom icon-root"
              role="presentation"
            >
              <use
                xlink:href="#caret-bottom"
              />
            </svg>
          </span>
        </button>
      </div>
      <div
        class="pagination-results"
      >
        Showing 1 to 1 of 10 entries.
      </div>
      <ul
        class="pagination pagination-root"
      >
        <li
          class="page-item disabled"
        >
          <a
            class="button-root page-link btn btn-unstyled"
            disabled=""
            href="/?page=0"
          >
            <svg
              class="lexicon-icon lexicon-icon-angle-left icon-root"
              role="presentation"
            >
              <use
                xlink:href="#angle-left"
              />
            </svg>
          </a>
        </li>
        <li
          class="page-item active"
        >
          <a
            class="button-root page-link btn btn-unstyled"
            href="/?page=1"
          >
            1
          </a>
        </li>
        <li
          class="page-item"
        >
          <a
            class="button-root page-link btn btn-unstyled"
            href="/?page=2"
          >
            2
          </a>
        </li>
        <li
          class="page-item"
        >
          <a
            class="button-root page-link btn btn-unstyled"
            href="/?page=3"
          >
            3
          </a>
        </li>
        <li
          class="page-item"
        >
          <div
            aria-haspopup="true"
            class="dropdown dropdown-root"
          >
            <button
              class="button-root dropdown-toggle page-link btn btn-unstyled"
              type="button"
            >
              <span
                class="text-truncate"
              >
                ...
              </span>
              
            </button>
          </div>
        </li>
        <li
          class="page-item"
        >
          <a
            class="button-root page-link btn btn-unstyled"
            href="/?page=10"
          >
            10
          </a>
        </li>
        <li
          class="page-item"
        >
          <a
            class="button-root page-link btn btn-unstyled"
            href="/?page=2"
          >
            <svg
              class="lexicon-icon lexicon-icon-angle-right icon-root"
              role="presentation"
            >
              <use
                xlink:href="#angle-right"
              />
            </svg>
          </a>
        </li>
      </ul>
    </div>
  </nav>
</div>
`;
